src/pages/gas/GasForm.tsx

```diff
@@ -0,0 +1,117 @@
+import { StateUpdater } from 'preact/hooks';
+import { useForm } from 'react-hook-form';
+
+import { OrmlTraitsAssetRegistryAssetMetadata } from '../../hooks/useBuyout/types';
+import { BlockchainAsset } from '../../components/Selector/AssetSelector/helpers';
+import From, { FromPropsWithVariant, FromVariants } from '../../components/Form/From';
+import { FeeHint } from './FeeHint';
+import { SubmitButton } from './SubmitButton';
+import { formatToSignificantDecimals } from './helpers';
+
+export type IssueFormValues = {
+  fromAmount: number;
+  toAmount: number;
+};
+
+interface GasFormProps {
+  onSubmit: (data: IssueFormValues) => void;
+  currencies: OrmlTraitsAssetRegistryAssetMetadata[];
+  setSelectedFromToken: StateUpdater<BlockchainAsset | undefined>;
+  selectedFromToken?: BlockchainAsset;
+  nativeCurrency: OrmlTraitsAssetRegistryAssetMetadata;
+  calcMin: () => number;
+  calcMax: () => number;
+  calcTo: (n: number) => number;
+  submissionPending: boolean;
+}
+
+export const GasForm: React.FC<GasFormProps> = ({
+  onSubmit,
+  currencies,
+  setSelectedFromToken,
+  selectedFromToken,
+  nativeCurrency,
+  calcMin,
+  calcMax,
+  calcTo,
+  submissionPending,
+}) => {
+  const { handleSubmit, register, setValue, watch, formState, clearErrors } = useForm<IssueFormValues>({});
+
+  const registerFromAmount = register('fromAmount', {
```

Comment:
I already mentioned it elsewhere but we need to make sure the validation triggers earlier, already when typing.